### ISSUE TITLE
910 Size Off

### ISSUE THREAD

**Issue by Teh-Bobo** [AF-VCD/pdf-bullets#38]
When I type the following bullet into a 910 it fits exactly; however, the tool puts the "std" onto the next line.

- Led 42 mbrs/3 AFSCs thru JQS update; certified mbrs/trainers on 133 rqmts--propelled nascent flight trng to wing std

Ignoring the quality of the bullet, it seems to fit the 910 with no space at all, the sizing in the tool may be slightly off.

**Comment by ckhordiasma** [comment]
related to https://github.com/AF-VCD/pdf-bullets/issues/34

**Comment by Teh-Bobo** [comment]
I see. Setting the size to 202.473mm was enough to make it fit. I see that the other bug is older, do you know if its an issue with how the text is rendered or if the wrong sizes are chosen for the boxes?

**Comment by ckhordiasma** [comment]
it's an issue with minor discrepancies in how the text is rendered in the
browser and in the PDF. In the PDF form internals, an XML format is used to
generate the text input boxes, with a width in mm specified. You can see
this if you go to http://brendandahl.github.io/pdf.js.utils/browser/ ,
upload an AF form, and navigate to Root -> AcroForm -> XFA -> download
stream 5 (the one after "template")

when you open up that file blob, search for `<field name="IIIComments" `
and you can see that x/y coordinates, as well as width/height for the form
field, is specified all in mm.

So on the website, I'm using CSS to specify the form width, and most of the
times it's consistent with the PDF form field width. However, it seems like
the PDF field rendering is more "willing" to cram characters in one line,
whereas in HTML rendering it will just decide to push the extra word off to
a new line.

In order to really fix this, I have to create my own text measurement
function (thinking of using text rendered in canvas for this) and then
manually insert line breaks based on the measurements.

**Comment by ckhordiasma** [comment]
Check out ckhordiasma.GitHub.io/pdf-bullets now, should work better now hopefully.

**Comment by Teh-Bobo** [comment]
Seems to work with every bullet I tried, did you end up doing your own rendering?

**Comment by ckhordiasma** [comment]
Yeah for each line I’m rendering a canvas 2D context (hidden) and running
context.measureText() to get the width in pixels.

Also making the assumption that the Adobe pdf form renders in 60 pixels per
inch. So far assumption seems correct.

Problem with using canvas (was a problem with my previous render method as
well) is that it’s hard to unit test for it.. the canvas library on npm
doesn’t implement measureText() very accurately.

So for unit testing, I ended up creating a dummy “mock” function that uses
a custom dictionary mapping between all possible UTF-8 characters and their
widths in times new Roman, 12 point font.

That ended up working so well that I think I’m going to just use that
function instead of the canvas method! (Haven’t pushed to repo yet).

 I have a script that generates the mapping for me, so I can implement in
different fonts and sizes if needed.

**Comment by Teh-Bobo** [comment]
Is that affected by things like kerning? I'm getting out of my depth a little here, but I know that certain fonts can adjust the spacing between letters depending on what letters surround them.

**Comment by ckhordiasma** [comment]
So the reason I chose HTML rendering, and then subsequently canvas
rendering, was that I was worried about things like the spacing between the
characters being variable.  However, based on my unit testing, either
kerning is negligible (like, less than a hundredth of a pixel), or there is
no kerning at all on this font.

I’ll need to do some googling to see if there is kerning on times new Roman
haha

**Comment by ckhordiasma** [comment]
Thanks for pointing me in the right direction about kerning! I was able to verify that the adobe font does NOT in fact implement kerning by default, so I disabled kerning accordingly on the website as well. I had to make some tweaks after that as well, but overall I think it should be more consistent!

**Comment by ckhordiasma** [comment]
```
- Led 42 mbrs/3 AFSCs thru JQS update; certified mbrs/trainers on 133 rqmts--propelled nascent flight trng to wing std
```

implemented fix in 060fc135f960bac78b7bdcad292e3b379bf137b9, also added above bullet to unit testing cases
